Implement whitened parametrisation (#71)
JuliaGaussianProcesses/ApproximateGPs.jl: docs/src/userguide.md

M = 15  # The number of inducing points
z = x[1:M]
```
The inducing inputs `z` imply some latent function values `u = f(z)`, sometimes called pseudo-points. The `SparseVariationalApproximation` specifies a distribution `q(u)` over the pseudo-points. In the case of GP regression, the optimal form for `q(u)` is a multivariate Gaussian, which is the only form of `q` currently supported by this package.
```julia
using Distributions, LinearAlgebra
q = MvNormal(zeros(length(z)), I)
```
Finally, we pass our `q` along with the inputs `f(z)` to obtain an approximate posterior GP:
```julia
fz = f(z, 1e-6)  # 'observe' the process at z with some jitter for numerical stability 

> approx = SparseVariationalApproximation(fz, q)  # Instantiate everything needed for the approximation

Reviewer: Hm, I guess this was missed in #70 because it's not actually a doctest... is that something that could be (reasonably) changed ?

Author: I don't know whether it's possible to add doctests to the docs directly (I think they have to be done in the docstring), but I might be wrong. There's definitely a case to be made that this entire doc should be a Literate script that we run every time we build the docs though.

Author: I've opened an issue about this as I'd rather do it as a follow up.

https://github.com/JuliaGaussianProcesses/ApproximateGPs.jl/issues/72

sva_posterior = posterior(approx)  # Create the approximate posterior
```

## The Evidence Lower Bound (ELBO)
The approximate posterior constructed above will be a very poor approximation, since `q` was simply chosen to have zero mean and covariance `I`. A measure of the quality of the approximation is given by the ELBO. Optimising this term with respect to the parameters of `q` and the inducing input locations `z` will improve the approximation.
```julia
elbo(SparseVariationalApproximation(fz, q), fx, y)
```
A detailed example of how to carry out such optimisation is given in [Regression: Sparse Variational Gaussian Process for Stochastic Optimisation with Flux.jl](@ref). For an example of non-conjugate inference, see [Classification: Sparse Variational Approximation for Non-Conjugate Likelihoods with Optim's L-BFGS](@ref).

> ## An Improved Parametrisation

Reviewer: It's not strictly an improvement (otherwise we shouldn't bother even implementing the non-whitened version!), and I've come across some work comparing centred vs non-centred representation (e.g. http://proceedings.mlr.press/v119/gorinova20a, even if in a slightly different context) that would make me wary to say "always whiten" - even though I do think it'd be a good default!

So this is one reason why I'd prefer the two parametrisations to be the same type - then we can make it easy for people to choose either, but for the default (when you don't provide it explicitly) to be whitened.

Author: > It's not strictly an improvement (otherwise we shouldn't bother even implementing the non-whitened version!), and I've come across some work comparing centred vs non-centred representation (e.g. http://proceedings.mlr.press/v119/gorinova20a, even if in a slightly different context) that would make me wary to say "always whiten" - even though I do think it'd be a good default!

Excellent point -- I'll add a link to that paper in the docs and soften the language.

> So this is one reason why I'd prefer the two parametrisations to be the same type - then we can make it easy for people to choose either, but for the default (when you don't provide it explicitly) to be whitened.

I see. I think we could implement this in a couple of ways, but I agree that using a type parameter is probably the most straightforward way for now. Will refactor and see what it looks like.

Author: I think I've addressed this, so will resolve.

M = 15  # The number of inducing points
z = x[1:M]
```

> The inducing inputs `z` imply some latent function values `u = f(z)`, sometimes called pseudo-points. The `SparseVariationalApproximation` specifies a distribution `q(u)` over the pseudo-points. In the case of GP regression, the optimal form for `q(u)` is a multivariate Gaussian, which is the only form of `q` currently supported by this package.

Reviewer 2: ```suggestion
The inducing inputs `z` imply some latent function values `u = f(z)`, sometimes called pseudo-points. The [`SparseVariationalApproximation`](@ref) specifies a distribution `q(u)` over the pseudo-points. In the case of GP regression, the optimal form for `q(u)` is a multivariate Gaussian, which is the only form of `q` currently supported by this package.
```

q = MvNormal(zeros(length(z)), I)
```
Finally, we pass our `q` along with the inputs `f(z)` to obtain an approximate posterior GP:
```julia
fz = f(z, 1e-6)  # 'observe' the process at z with some jitter for numerical stability 
approx = SparseVariationalApproximation(fz, q)  # Instantiate everything needed for the approximation

sva_posterior = posterior(approx)  # Create the approximate posterior
```

## The Evidence Lower Bound (ELBO)
The approximate posterior constructed above will be a very poor approximation, since `q` was simply chosen to have zero mean and covariance `I`. A measure of the quality of the approximation is given by the ELBO. Optimising this term with respect to the parameters of `q` and the inducing input locations `z` will improve the approximation.
```julia
elbo(SparseVariationalApproximation(fz, q), fx, y)
```
A detailed example of how to carry out such optimisation is given in [Regression: Sparse Variational Gaussian Process for Stochastic Optimisation with Flux.jl](@ref). For an example of non-conjugate inference, see [Classification: Sparse Variational Approximation for Non-Conjugate Likelihoods with Optim's L-BFGS](@ref).

# Available Parametrizations

> Two parametrizations of `q(u)` are presently available: `Centered` and `NonCentered`.

Reviewer 2: ```suggestion
Two parametrizations of `q(u)` are presently available: [`Centered`](@ref) and [`NonCentered`](@ref).
```

The `Centered` parametrization expresses `q(u)` directly in terms of its mean and covariance.
The `NonCentered` parametrization instead parametrizes the mean and covariance of

> `ε := cholesky(cov(u)).U' \ (u - mean(u))`.

Reviewer 2: ```suggestion
`ε := cholesky(cov(u)).U' \ (u - mean(u))`.
These parametrizations are also known respectively as "Unwhitened" and "Whitened".
```

Reviewer: ```suggestion
`ε := L \ (u - mean(u))`, where `L` is the lower-triangular Cholesky factor of `cov(u)`.
```
to make it a bit clearer/easier to read? it's only pseudo-code anyways

The choice of parametrization can have a substantial impact on the time it takes for ELBO
optimisation to converge, and which parametrization is better in a particular situation is
not generally obvious.

> That being said, the `NonCentered` parametrization is often superior, so it is the default --

Reviewer 2: "superior" is a bit vague no? Maybe something like 
```suggestion
That being said, the `NonCentered` parametrization has better convergence properties, so it is the default --
```

Author: ```suggestion
That being said, the `NonCentered` parametrization often converges in fewer iterations, so it is the default --
```
I've softened a little, because it's not always the case that it converges more quickly, just a lot of the time.

```julia
elbo(SparseVariationalApproximation(fz, q), fx, y)
```
A detailed example of how to carry out such optimisation is given in [Regression: Sparse Variational Gaussian Process for Stochastic Optimisation with Flux.jl](@ref). For an example of non-conjugate inference, see [Classification: Sparse Variational Approximation for Non-Conjugate Likelihoods with Optim's L-BFGS](@ref).

# Available Parametrizations

Two parametrizations of `q(u)` are presently available: [`Centered`](@ref) and [`NonCentered`](@ref).
The `Centered` parametrization expresses `q(u)` directly in terms of its mean and covariance.
The `NonCentered` parametrization instead parametrizes the mean and covariance of
`ε := cholesky(cov(u)).U' \ (u - mean(u))`.
These parametrizations are also known respectively as "Unwhitened" and "Whitened".

The choice of parametrization can have a substantial impact on the time it takes for ELBO

> optimisation to converge, and which parametrization is better in a particular situation is

Reviewer: ```suggestion
optimization to converge, and which parametrization is better in a particular situation is
```
:stuck_out_tongue:

not generally obvious.
That being said, the `NonCentered` parametrization often converges in fewer iterations, so it is the default --
it is what is used in all of the examples above.

If you require a particular parametrization, simply use the 3-argument version of the
approximation constructor:
```julia
SparseVariationalApproximation(Centered(), fz, q)
SparseVariationalApproximation(NonCentered(), fz, q)
```

> For a discussion around these two parametrizations, see e.g. [^Gorinova]

Reviewer: One aspect that Gorinova et al are perhaps not discussing specifically is that, when considering GPs, there's a lot of correlation imposed by the kernel on `u = f(Z)`, section 3.4 in https://www.stat.berkeley.edu/~paciorek/diss/paciorek-thesis.pdf discusses that more concretely for the GP case, though I'm not sure if it's the _best_ reference on this topic, maybe worth including as well?
